### DISCUSSION TITLE
NextJs rewrites to support Incremental adoption

### DISCUSSION THREAD

**Discussion by iamsharzil** [vercel/next.js#37007] [Help]
I've two projects,
- **Project 1** - React + Electrode - *www.example.com/au*
- **Project 2** - Next Js - *www.example.com/au/sell-your-car*

**Project 2** has a middleware to detect whether the **request** is coming from **desktop/mobile**.

The **Project 2** page folder looks like the following:

 - pages
   - _viewport
     - desktop
       - au 
         - sell-your-car
     - mobile
       - au 
         - sell-your-car
    

> _middleware.tsx

    function getViewport(userAgent: UserAgent | null | undefined) {
      if (!userAgent?.ua) return;
      const isMobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(userAgent.ua);
      return isMobile ? "mobile" : "desktop";
    }
    
    /** @type {import("next/server").NextMiddleware} */
    
    export function middleware(req: NextRequest) {
      const url = req.nextUrl.clone();
    
      // Prevent internals from being accessed canonically
      // (i.e. if you try to go to example.com/_viewport it should give a 404)
      if (url.pathname.startsWith(`/_viewport`)) {
        return new Response(null, { status: 404 });
      }
    
      const viewport = getViewport(req.ua);
    
      url.pathname = `_viewport/${viewport}${url.pathname}`;
    
      return NextResponse.rewrite(url);
    }

I want to set up rewrites in such a way that they should support all the routes after the **/sell-your-car** route.

e.g.
 - *www.example.com/au/sell-your-car*  
 - *www.example.com/au/sell-your-car/1*
 - *www.example.com/au/sell-your-car/1/2* (and so on)

Other than that all the routes should redirect to the Project 1 (React)

> next.config.js

    async rewrites() {
        return [
          {
            source: "/:path*",
            destination: "/:path*"
          },
          {
            source: "/:path*",
            destination: "http://localhost:3001/:path*" // Rewrite non-migrated routes to Create React App
          }
        ];
      }

With the current config, the **/sell-your-car** page is getting refreshed continuously.

**Comment by icyJoseph** [comment]
> With the current config, the /sell-your-car page is getting refreshed continuously.

Could you be more specific? What happens to make the page refresh? User navigation?

**Reply by iamsharzil** [reply]
@icyJoseph Thanks for the quick response.

The thing is that I can't even navigate to any other page. The page is getting refreshed continuously. I'll attach a video that will explain it clearly.

https://user-images.githubusercontent.com/49302938/168996840-b53cf2c0-26ba-4bf2-98f2-b6d71857791f.mov

.

**Reply by icyJoseph** [reply]
Oh wow, I was able to reproduce this. Let me take a look over lunch, looks very shady!

Btw, if you are already using `middleware`, why not handle all redirects/rewrites there?

**Reply by iamsharzil** [reply]
@icyJoseph I thought using rewrites is a clean way of doing this. I'll try with the middleware as well but it would be very helpful if you could share the solution. I've also asked on Discord, Stack overflow, and Reddit but didn't get the answer. 😅

**Reply by icyJoseph** [reply]
Alright, there seems to be some collision between the `middleware` rewrite, and the `next.config.js` rewrites. On my reproduction demo I have `/_viewport/desktop/sell.tsx` as only path.

One interesting thing is that the middleware runs even on the CRA app request! I had not thought about this.

I am still trying to get an accurate model of what's going on, but I think it has to be:

1. First request arrives and it is middleware rewritten, and served to the client
2. `next.config.js` rewrite finds no matching file, tries to rewrite to CRA, which triggers, an expected, client side page load
3. Page re-loads on /au/sell-your-car, which brings us back to 1

At least the observations I have made tell me this much. I see the `x-middleware-rewrite` present on the response headers on the client. 

One strange thing is that, if I define, `pages/sell.tsx`, and put whatever there, **then the infinite reloading stops**, and it shows the contents of `/_viewport/desktop/sell.tsx` - odd stuff...

With only middleware, I guess it'd be a bit more cumbersome, but you'd basically define which paths exist on your app, for example `sell`, and if the middleware pathname is that, rewrite it to the file you want, otherwise, 
 
To get it to work with middleware, do this, but it will only work in Production Mode for the CRA/Vite app (webpack development scripts are a bit unfriendly in this scenario):

```ts
import { NextResponse, type NextRequest } from "next/server";

/** @type {import("next/server").NextMiddleware} */
export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  )
    return NextResponse.next();

  if (pathname.startsWith(`/_viewport`)) {
    return new Response(null, { status: 404 });
  }

  // some special path that needs to show something conditionally
  if (pathname === "/sell") {
    req.nextUrl.pathname = `/_viewport/desktop${pathname}`;
    const response = NextResponse.rewrite(req.nextUrl);

    return response;
  }

  if (pathname === "/" || pathname === "/blog") {
    // we have these defined, let them pass
    return NextResponse.next();
  }

  // rewrite to the CRA app everything else
  return NextResponse.rewrite(`http://localhost:3001${pathname}`);
}
```

I am not sure this is 100% bullet proof though, but you could do this middleware and test, or define `pages/au/sell-your-car.tsx` with just a dummy component inside, or migrate everything to Next.js :)

**Reply by iamsharzil** [reply]
@icyJoseph Thank you for sharing the code. I just realized that rewrite is also getting triggered in middleware. So, I'll have to make the changes in middleware instead of the next config. I'll work around it and share the logic over here. 
Thanks a lot for your time. 🙂
